### ISSUE TITLE
Make LSP server shutdown configurable

### ISSUE THREAD

**Issue by kaikreuzer** [eclipse/xtext-core#885]
We (that is the Eclipse SmartHome project) recently upgraded to Xtext 2.14 and now ran into the issue that [our server is silently shutting down](https://github.com/eclipse/smarthome/issues/6291), if VS Code is disconnecting after having used LSP.
We tracked that down to https://github.com/eclipse/xtext-core/pull/474 and being a "feature" of an LSP server. While I see that this is a part of the specification and thus has to be supported, it is nonetheless an absolutely no-go to call `System.exit()` from within some library, especially within an OSGi container.
I guess our situation, where LSP is run as a service on an existing server (and not as a JVM process on its own), clearly shows that this is a pretty nasty behavior.

Imho, the clean way to implement this in an OSGi way would be to provide a hook for listening to exit requests and have a default implementation of such a listener, which cleanly shuts down the OSGi container, while giving consumers the chance to deactivate/replace this default implementation (e.g. through config admin). Wdyt?

**Comment by kthoms** [comment]
I understand your concern, but as you already found out it is a feature. Your requirement is to shut down the OSGi container then, and a listener could be a way to do so. However, the Xtext core bundles are not aware of OSGi and don't have a dependency on that. And I do not think that one should be introduced. This means it must be somehow added from outside.

I could imagine to delegate the behaviour to a shutdown handle that could be overridden by a specific language. Or do you have an idea how to shut down the OSGi container without a dependency here?

**Comment by kaikreuzer** [comment]
> Your requirement is to shut down the OSGi container then

No, not at all - at most, the LSP server bundles should stop their services and must not impact any other code running on the OSGi container. Btw, just for my understanding, what is the background of this "feature"? If the server is stopped upon closing the editor, how is it started when opening the editor again?

> This means it must be somehow added from outside.

Well, the concept of registering listeners (`addExitNotificationListener` method) isn't related to OSGi, so this "logic" can easily be drawn out of the code code. And the default listener implementation could e.g. be deactivated by a system property (to not have any OSGi dependency either).

Btw, in any case, the code should properly handle Security exceptions. When using a Java security manager, you currently get something like this in your log:
```
Oct 01, 2018 9:25:16 AM org.eclipse.lsp4j.jsonrpc.RemoteEndpoint handleNotification
WARNING: Notification threw an exception: {
  "jsonrpc": "2.0",
  "method": "exit",
  "params": null
}
java.lang.RuntimeException: java.lang.reflect.InvocationTargetException
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:63)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.notify(GenericEndpoint.java:148)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.handleNotification(RemoteEndpoint.java:216)
	at org.eclipse.lsp4j.jsonrpc.RemoteEndpoint.consume(RemoteEndpoint.java:183)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.handleMessage(StreamMessageProducer.java:188)
	at org.eclipse.lsp4j.jsonrpc.json.StreamMessageProducer.listen(StreamMessageProducer.java:90)
	at org.eclipse.lsp4j.jsonrpc.json.ConcurrentMessageProcessor.run(ConcurrentMessageProcessor.java:95)
	at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:511)
	at java.util.concurrent.FutureTask.run(FutureTask.java:266)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1142)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:617)
	at java.lang.Thread.run(Thread.java:745)
Caused by: java.lang.reflect.InvocationTargetException
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.eclipse.lsp4j.jsonrpc.services.GenericEndpoint.lambda$null$0(GenericEndpoint.java:61)
	... 11 more
Caused by: java.lang.SecurityException: Prevented System.exit() call.
	at org.openhab.core.internal.CoreActivator$1.checkPermission(CoreActivator.java:51)
	at java.lang.SecurityManager.checkExit(SecurityManager.java:761)
	at java.lang.Runtime.exit(Runtime.java:107)
	at java.lang.System.exit(System.java:971)
	at org.eclipse.xtext.ide.server.LanguageServerImpl.exit(LanguageServerImpl.java:334)
	... 16 more
```
(just to repeat my statement from above: " it is an absolute no-go to call System.exit() from within some library".)

**Comment by cdietrich** [comment]
@svenefftinge please comment what you think regarding this issue. Adding a new listener is bad since it breaks api. What about a new callback that defaults to calling system.exit

**Comment by kaikreuzer** [comment]
> the code should properly handle Security exceptions

FTR, the server might be in some limbo state if the security manager prevented the system exit as the LSP server might remain in "being shut down" state. (I couldn't clearly reproduce this, see [this comment](https://github.com/eclipse/smarthome/issues/6291#issuecomment-425843106).)

**Comment by svenefftinge** [comment]
I agree with kai in that we should not call System.exit down in a library. Instead we should provide a onExit callback that is provided from the starter code.

**Comment by cdietrich** [comment]
Fixed in 2.16
